python/llm/test/inference_gpu/test_transformers_api.py

```diff
@@ -36,38 +37,192 @@
+            logits_optimized_model = (opt_model(input_ids)).logits
+            # the list `layer_output` has only one element.
+            opt_layer_tensor = self.layer_outputs[0]
+            opt_model.to('cpu')
+
+            attn_output_diff = []
+            for i, (t1, t2) in enumerate(zip(layer_tensor, opt_layer_tensor)):
+                if t1 is not None and t2 is not None:
+                    if isinstance(t1, torch.Tensor) and isinstance(t2, torch.Tensor):
+                        # 'attn_output' is of type torch.Tensor.
+                        attn_output_diff.append(t1 - t2)
+                    else:
+                        # 'past_key_value'is of type tuple as default.
+                        for i, (t3, t4) in enumerate(zip(t1, t2)):
+                            if model.config.architectures[0] == "ChatGLMModel" and \
+                                    hasattr(model.config, 'padded_vocab_size') and \
+                                    model.config.padded_vocab_size == 65024:
+                                # chatglm2's past_key_value is expanded 16x for some speedup.
+                                # We need to narrow it here.
+                                t4 = t4[:, :, 15:17, :]
+                            attn_output_diff.append(t3 - t4)
+
+            max_diff_tensor = [torch.max(item).item() for item in attn_output_diff]
+            print(max_diff_tensor)
+            assert all(max_diff <= lower_bound for max_diff in max_diff_tensor)
+
+
+    def test_falcon_gpu_model(self):
+
+        Model = AutoModelForCausalLM
+        Tokenizer = AutoTokenizer
+        model_path = os.environ.get('FALCON_7B_ORIGIN_PATH')
+        # currently only compare the output of the last self-attention layer.
+        layer_norm = "transformer.h.31.input_layernorm"
+        self_attn = "transformer.h.31.self_attention"
+        lower_bound = 0
+
+        self.run_optimize_gpu_model(Model, Tokenizer, model_path, self_attn, layer_norm, lower_bound)
+
+
+    def test_llama_gpu_model(self):
+
+        Model = AutoModelForCausalLM
+        Tokenizer = AutoTokenizer
+        model_path = os.environ.get('LLAMA2_7B_ORIGIN_PATH')
+        # currently only compare the output of the last self-attention layer.
+        layer_norm = "model.layers.31.input_layernorm"
+        self_attn = "model.layers.31.self_attn"
+        lower_bound = 5e-2
+
+        self.run_optimize_gpu_model(Model, Tokenizer, model_path, self_attn, layer_norm, lower_bound)
+
+    def test_chatglm2_gpu_model(self):
+
+        Model = AutoModel
+        Tokenizer = AutoTokenizer
+        model_path = os.environ.get('CHATGLM2_6B_ORIGIN_PATH')
+        # currently only need to compare the output of one self-attention layer.
+        layer_norm = "transformer.encoder.layers.27.input_layernorm"
+        self_attn = "transformer.encoder.layers.27.self_attention"
+        lower_bound = 1e-3
```

Comment:
The main thing we need to figure out is why previously 5e-5 works but now it doesn't work. Maybe to relax the bound is correct, but we need to confirm the reason.

Reply:
More information of this difference can be found here: https://github.com/analytics-zoo/nano/issues/852#issuecomment-1863733041

We may have some more investigation later but I think we can merge this PR for now :)